indico/modules/events/layout/controllers/menu.py

```diff
@@ -247,4 +248,8 @@ def _process_args(self):
         self.page = EventPage.get_or_404(request.view_args['page_id'])
 
     def _process(self):
+        registered_only = self.page.menu_entry.registered_only
+        registered_users = [x.user for x in get_registered_event_persons(self.event)]
+        if registered_only and session.user not in registered_users and not self.event.can_manage(session.user):
+            raise Forbidden
```

Comment:
This is an access check so it should be in `_check_access`.

`get_registered_event_persons` is not the best choice here - you do not really care about EventPersons here but just if the user is registered. Querying all registered users is also expensive, you only need to know about the current user here.

I'd suggest adding this method to `User` to a cheap check:

```python
def is_registered_in_event(self, event):
    """Check whether the user is registered in the event.
    
    This takes both unpaid and complete registrations into account.
    """
    from indico.modules.events.registration.models.registrations import Registration, RegistrationState
    return (Registration.query.with_parent(self)
            .join(Registration.registration_form)
            .filter(Registration.event == event,
                    Registration.state.in_([RegistrationState.unpaid, RegistrationState.complete]),
                    ~Registration.is_deleted,
                    ~RegistrationForm.is_deleted)
            .has_rows())
```